src/disjoint_mut.rs

```diff
@@ -687,30 +696,35 @@ where
 {
     type Output = <[T] as Index<Range<usize>>>::Output;
 
+    #[inline(always)] // Need to inline to see bounds checks in order to potentially elide them.
     #[cfg_attr(debug_assertions, track_caller)]
     unsafe fn get_mut(self, slice: *mut [T]) -> *mut Self::Output {
         // SAFETY: The safety precondition for this trait method
         // requires that we can immutably dereference `slice`.
         let len = unsafe { (*slice).len() };
-        let Range { start, end } = self.clone().into().range;
-        let end = if end == usize::MAX { len } else { end };
-        if start <= end && start < len && end <= len {
+        let Range { start, end } = self.to_range(len);
+        if start <= end && end <= len {
             // SAFETY: We have checked that `start` is less than the
             // allocation length therefore cannot overflow. `slice` is a
             // valid pointer into an allocation of sufficient length.
             let data = unsafe { (slice as *mut T).add(start) };
             ptr::slice_from_raw_parts_mut(data, end - start)
         } else {
-            if start > end {
-                panic!("slice index starts at {start} but ends at {end}");
-            }
-            if end > len {
-                panic!("range end index {end} out of range for slice of length {len}");
-            }
-            if start >= len {
-                panic!("range start index {start} out of range for slice of length {len}")
+            #[inline(never)]
+            #[cfg_attr(debug_assertions, track_caller)]
+            fn out_of_bounds(start: usize, end: usize, len: usize) -> Infallible {
+                if start > end {
+                    panic!("slice index starts at {start} but ends at {end}");
+                }
+                if end > len {
+                    panic!("range end index {end} out of range for slice of length {len}");
+                }
+                if start >= len {
+                    panic!("range start index {start} out of range for slice of length {len}")
+                }
```

Comment:
You took this case out? (zero-length slice) That's fine, but do we need this panic then?

Reply:
I don't think so?  I wasn't paying attention to this much since it's in the error path, but it does seem like this one is unreachable now, right?